#include "mainwindow.h"
#include "ui_mainwindow.h"
#include "tcpclient.h"
#include <QMessageBox>
#include <QLineEdit>
#include <QPushButton>
#include <QTextBrowser>
#include <QTimer>
#include <exception>
#include <sstream>

MainWindow::MainWindow(const QString& login, const QString& nickname, QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
    , currentUserLogin(login)
    , currentUserNickname(nickname)
    , lastChatContent("")
{
    ui->setupUi(this);
    setWindowTitle("Ð§Ð°Ñ‚ â€” " + nickname);

    // Ð¢Ð°Ð¹Ð¼ÐµÑ€ Ð°Ð²Ñ‚Ð¾Ð¾Ð±Ð½Ð¾Ð²Ð»ÐµÐ½Ð¸Ñ
    chatUpdateTimer = new QTimer(this);
    connect(chatUpdateTimer, &QTimer::timeout, this, [this]() {
        if (isPublicChat) {
            loadChatHistory();
        } else {
            loadPrivateChat();
        }
    });
    chatUpdateTimer->start(3000);

    // Ð¢Ð°Ð¹Ð¼ÐµÑ€ Ð¿Ñ€Ð¾Ð²ÐµÑ€ÐºÐ¸ ÑÑ‚Ð°Ñ‚ÑƒÑÐ°
    statusCheckTimer = new QTimer(this);
    connect(statusCheckTimer, &QTimer::timeout, [this]() {
        try {
            TcpClient client;
            std::string cmd = "check_status " + currentUserLogin.toStdString();
            std::string resp = client.sendCommand(cmd);

            if (resp.find("STATUS:BANNED") != std::string::npos) {
                QMessageBox::critical(this, "Ð”Ð¾ÑÑ‚ÑƒÐ¿ Ð·Ð°Ð¿Ñ€ÐµÑ‰Ñ‘Ð½",
                                      "Ð’Ð°Ñˆ Ð°ÐºÐºÐ°ÑƒÐ½Ñ‚ Ð·Ð°Ð±Ð»Ð¾ÐºÐ¸Ñ€Ð¾Ð²Ð°Ð½. ÐžÐ±Ñ€Ð°Ñ‚Ð¸Ñ‚ÐµÑÑŒ Ðº Ð°Ð´Ð¼Ð¸Ð½Ð¸ÑÑ‚Ñ€Ð°Ñ‚Ð¾Ñ€Ñƒ.");
                this->close();
            }
            else if (resp.find("STATUS:KICKED") != std::string::npos) {
                QMessageBox::warning(this, "ÐžÑ‚ÐºÐ»ÑŽÑ‡ÐµÐ½Ð¾",
                                     "Ð’Ñ‹ Ð±Ñ‹Ð»Ð¸ Ð¾Ñ‚ÐºÐ»ÑŽÑ‡ÐµÐ½Ñ‹ Ð¾Ñ‚ ÑÐµÑ€Ð²ÐµÑ€Ð°. ÐŸÐ¾Ð¶Ð°Ð»ÑƒÐ¹ÑÑ‚Ð°, Ð²Ð¾Ð¹Ð´Ð¸Ñ‚Ðµ ÑÐ½Ð¾Ð²Ð°.");
                this->close();
            }
        } catch (...) {
            // Ð˜Ð³Ð½Ð¾Ñ€
        }
    });
    statusCheckTimer->start(3000);

    // Ð¡Ð¸Ð³Ð½Ð°Ð»Ñ‹
    connect(ui->sendButton, &QPushButton::clicked, this, &MainWindow::onSendClicked);
    connect(ui->messageEdit, &QLineEdit::returnPressed, this, &MainWindow::onSendClicked);
    connect(ui->logoutButton, &QPushButton::clicked, this, &MainWindow::onLogoutClicked);

    loadUserList();
    loadChatHistory();
}

MainWindow::~MainWindow()
{
    delete ui;
}

void MainWindow::onSendClicked()
{
    QString text = ui->messageEdit->text().trimmed();
    if (text.isEmpty()) return;

    try {
        TcpClient client;
        std::string cmd;
        if (isPublicChat) {
            cmd = "send " + currentUserLogin.toStdString() + " ALL " + text.toStdString();
        } else {
            cmd = "send " + currentUserLogin.toStdString() + " " + currentChatTarget.toStdString() + " " + text.toStdString();
        }

        std::string resp = client.sendCommand(cmd);
        if (resp.find("OK:") != std::string::npos) {
            ui->messageEdit->clear();
            if (isPublicChat) {
                loadChatHistory();
            } else {
                loadPrivateChat();
            }
        }
    } catch (...) {
        // Ð¾Ð±Ñ€Ð°Ð±Ð¾Ñ‚ÐºÐ° Ð¾ÑˆÐ¸Ð±ÐºÐ¸
    }
}

void MainWindow::onLogoutClicked()
{
    try {
        TcpClient client;
        std::string cmd = "logout " + currentUserLogin.toStdString();
        client.sendCommand(cmd);
    } catch (...) {
        // Ð¸Ð³Ð½Ð¾Ñ€
    }
    this->close();
}

void MainWindow::onChatSelected(QListWidgetItem* item)
{
    QString text = item->text();

    if (text == "ðŸ’¬ ÐžÐ±Ñ‰Ð¸Ð¹ Ñ‡Ð°Ñ‚") {
        isPublicChat = true;
        currentChatTarget.clear();
        loadChatHistory();
    } else {
        isPublicChat = false;
        // Ð£Ð±Ð¸Ñ€Ð°ÐµÐ¼ Ð¿Ñ€ÐµÑ„Ð¸ÐºÑ "ðŸ“© ", ÐµÑÐ»Ð¸ Ð¾Ð½ ÐµÑÑ‚ÑŒ
        if (text.startsWith("ðŸ“© ")) {
            currentChatTarget = text.mid(3);
        } else {
            currentChatTarget = text;
        }
        loadPrivateChat();
    }
}

void MainWindow::loadUserList()
{
    try {
        TcpClient client;
        std::string resp = client.sendCommand("list_users");

        size_t begin = resp.find("USERS_BEGIN\n");
        size_t end = resp.find("\nUSERS_END");

        ui->chatListWidget->clear();
        ui->chatListWidget->addItem("ðŸ’¬ ÐžÐ±Ñ‰Ð¸Ð¹ Ñ‡Ð°Ñ‚");

        // Ð”Ð¾Ð±Ð°Ð²Ð»ÑÐµÐ¼ Ð´Ñ€ÑƒÐ³Ð¸Ñ… Ð¿Ð¾Ð»ÑŒÐ·Ð¾Ð²Ð°Ñ‚ÐµÐ»ÐµÐ¹ (ÐºÑ€Ð¾Ð¼Ðµ ÑÐµÐ±Ñ)
        if (begin != std::string::npos && end != std::string::npos && end > begin) {
            size_t start = begin + 12;
            size_t len = end - start;
            std::string usersPart = resp.substr(start, len);

            std::istringstream iss(usersPart);
            std::string nick;
            while (std::getline(iss, nick)) {
                if (!nick.empty() && nick != currentUserNickname.toStdString()) {
                    ui->chatListWidget->addItem(QString::fromStdString(nick));
                }
            }
        }

        // Ð’Ð¡Ð•Ð“Ð”Ð Ð´Ð¾Ð±Ð°Ð²Ð»ÑÐµÐ¼ "Ñ‡Ð°Ñ‚ Ñ ÑÐ°Ð¼Ð¸Ð¼ ÑÐ¾Ð±Ð¾Ð¹" (Ð²Ð½Ðµ if!)
        ui->chatListWidget->addItem("ðŸ“© " + currentUserNickname);

        static bool connected = false;
        if (!connected) {
            connect(ui->chatListWidget, &QListWidget::itemClicked,
                    this, &MainWindow::onChatSelected);
            connected = true;
        }
    } catch (...) {
        ui->chatListWidget->addItem("âš ï¸ ÐÐµ ÑƒÐ´Ð°Ð»Ð¾ÑÑŒ Ð·Ð°Ð³Ñ€ÑƒÐ·Ð¸Ñ‚ÑŒ Ð¿Ð¾Ð»ÑŒÐ·Ð¾Ð²Ð°Ñ‚ÐµÐ»ÐµÐ¹");
    }
}

void MainWindow::loadChatHistory()
{
    try {
        TcpClient client;
        std::string resp = client.sendCommand("show_public " + currentUserLogin.toStdString());

        size_t beginPos = resp.find("MESSAGES_BEGIN\n");
        size_t endPos = resp.find("\nMESSAGES_END");

        QString content;
        if (beginPos != std::string::npos && endPos != std::string::npos && endPos > beginPos) {
            size_t start = beginPos + 14;
            size_t len = endPos - start;
            std::string messagesPart = resp.substr(start, len);
            content = QString::fromStdString(messagesPart);
        } else {
            content = "ÐžÐ±Ñ‰Ð¸Ð¹ Ñ‡Ð°Ñ‚ Ð¿ÑƒÑÑ‚.";
        }

        if (content != lastChatContent) {
            ui->chatTextEdit->setPlainText(content);
            lastChatContent = content;
        }
    } catch (const std::exception& e) {
        QString error = "ÐžÑˆÐ¸Ð±ÐºÐ° Ð·Ð°Ð³Ñ€ÑƒÐ·ÐºÐ¸ Ñ‡Ð°Ñ‚Ð°.";
        if (error != lastChatContent) {
            ui->chatTextEdit->setPlainText(error);
            lastChatContent = error;
        }
    }
}

void MainWindow::loadPrivateChat()
{
    try {
        TcpClient client;
        std::string cmd = "show_private "
                          + currentUserLogin.toStdString() + " "
                          + currentChatTarget.toStdString();
        std::string resp = client.sendCommand(cmd);

        size_t beginPos = resp.find("MESSAGES_BEGIN\n");
        size_t endPos = resp.find("\nMESSAGES_END");

        QString content;
        if (beginPos != std::string::npos && endPos != std::string::npos && endPos > beginPos) {
            size_t start = beginPos + 14;
            size_t len = endPos - start;
            std::string messagesPart = resp.substr(start, len);
            content = QString::fromStdString(messagesPart);
        } else {
            content = "ÐÐµÑ‚ Ð»Ð¸Ñ‡Ð½Ñ‹Ñ… ÑÐ¾Ð¾Ð±Ñ‰ÐµÐ½Ð¸Ð¹.";
        }

        if (content != lastChatContent) {
            ui->chatTextEdit->setPlainText(content);
            lastChatContent = content;
        }
    } catch (const std::exception& e) {
        QString error = "ÐžÑˆÐ¸Ð±ÐºÐ° Ð·Ð°Ð³Ñ€ÑƒÐ·ÐºÐ¸ Ð»Ð¸Ñ‡Ð½Ð¾Ð³Ð¾ Ñ‡Ð°Ñ‚Ð°.";
        if (error != lastChatContent) {
            ui->chatTextEdit->setPlainText(error);
            lastChatContent = error;
        }
    }
}
